### PR TITLE
feat(#major); Velodrome Finance; v1.0.0

### PR DESCRIPTION
**Pull request by l0c4t0r**
PR for new subgraph for Velodrome Finance on optimism.

Currently syncing at l0c4t0r/laboratory5

### CODE REVIEW

**Review by steegecs — COMMENTED** [review]
There are a lot of very larger TVL and volume errors. Also, when I look at the financial daily snapshots for TVL and daily volume in USD, there are times where the TVL looks normal, but the volume is way off - this is odd to me - though I am not sure if it requires an additional fix besides pricing. 

https://subgraphs.messari.io/subgraph?endpoint=l0c4t0r/laboratory5&tab=protocol

File: subgraphs/velodrome-finance/src/common/getters.ts [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+    protocol.type = ProtocolType.EXCHANGE;
+    protocol.totalValueLockedUSD = BIGDECIMAL_ZERO;
+    protocol.cumulativeVolumeUSD = BIGDECIMAL_ZERO;
+    protocol.cumulativeSupplySideRevenueUSD = BIGDECIMAL_ZERO;
+    protocol.cumulativeProtocolSideRevenueUSD = BIGDECIMAL_ZERO;
+    protocol.cumulativeTotalRevenueUSD = BIGDECIMAL_ZERO;
+    protocol.cumulativeUniqueUsers = INT_ZERO;
+    protocol.totalPoolCount = INT_ZERO;
+    protocol._stableFee = BigDecimal.fromString("0.02"); // Value hardcoded in factory constructor
+    protocol._volatileFee = BigDecimal.fromString("0.02"); // Value hardcoded in factory constructor
```
[the remark this answers is not in the document]
Comment (l0c4t0r) [reply subgraphs/velodrome-finance/src/common/getters.ts:284]: Both the stable fee and volatile fee are initialised at 0.02% on the pair factory, hence we can hardcode this into the getter.
However, thanks for noting this.  The protocol can update this fee via the setFee function.  I had forgotten to implement updating the fees for the pools when this happens, which I will do.

File: subgraphs/velodrome-finance/src/mappings/helpers/rewards.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    pool.rewardTokenEmissionsAmount = [BIGINT_ZERO];
+    pool.rewardTokenEmissionsUSD = [BIGDECIMAL_ZERO];
+    pool.save();
+}
+
+export function updateRewards(event: DistributeReward): void {
+    let gauge = getOrCreateGauge(event.params.gauge);
+
+    let pool = getLiquidityPool(Address.fromString(gauge.pool));
+    const rewardTokenEmissionsAmount = event.params.amount.div(BigInt.fromI32(7));
```
[the remark this answers is not in the document]
Comment (l0c4t0r) [reply subgraphs/velodrome-finance/src/mappings/helpers/rewards.ts:36]: The VELO rewards are distributed to the gauges weekly and emits DistributeRewards.
Hence I divide by 7 to get the reward amount per day.

File: subgraphs/velodrome-finance/src/common/metrics.ts [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+  let token1 = getOrCreateToken(Address.fromString(pool.inputTokens[1]));
+
+  let amount0USD = applyDecimals(event.params.amount0, token0.decimals).times(
+    token0.lastPriceUSD!
+  );
+  let amount1USD = applyDecimals(event.params.amount1, token1.decimals).times(
+    token1.lastPriceUSD!
+  );
+
+  let protocolSideRevenue = amount0USD.plus(amount1USD);
```
[the remark this answers is not in the document]
Comment (l0c4t0r) [reply subgraphs/velodrome-finance/src/common/metrics.ts:60]: > There are a lot of very larger TVL and volume errors. Also, when I look at the financial daily snapshots for TVL and daily volume in USD, there are times where the TVL looks normal, but the volume is way off - this is odd to me - though I am not sure if it requires an additional fix besides pricing.
> 
> https://subgraphs.messari.io/subgraph?endpoint=l0c4t0r/laboratory5&tab=protocol

Communicated with @steegecs on a call early. My fault for not including a link, the one that was reviewed was an older version that was synced.  The subgraph based on code in this PR is still syncing as the pending version of l0c4t0r/laboratory5.  The dashboard to QC should be:
https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/id/QmYTUWDYJapauCHZHR5FZiBk5Yte7KrzcMfjUFhuo4Zby1&tab=protocol

File: subgraphs/velodrome-finance/src/common/getters.ts [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+        ? BIGDECIMAL_ONE
+        : BIGDECIMAL_ZERO;
+    token.lastPriceBlockNumber = BIGINT_ZERO;
+    token.save();
+  }
+  return token;
+}
+
+export function getOrCreateRewardToken(address: Address): RewardToken {
+  let rewardToken = RewardToken.load(address.toHexString());
```
[the remark this answers is not in the document]
Comment (l0c4t0r) [reply subgraphs/velodrome-finance/src/common/getters.ts:55]: Got it - will fix

File: subgraphs/velodrome-finance/src/common/metrics.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (l0c4t0r) [reply subgraphs/velodrome-finance/src/common/metrics.ts:60]: Yes so this is how it works:
1. User provides tokens to a specific pool (LP) and get LP tokens
2. Swap fees 100% go to protocol
3. User stake LP tokens to get VELO emissions.  *This is the only "yield" directly going to LPs*
4. VELO can be locked for veVELO.  Similar to curve, longer you lock, the more veVELO you get. 
5. veVELO holders get rewarded with protocol fees (collected in 2.), bribes, rebases and governance power
6. veVELO holders can vote on pools/gauges to receive VELO emissions that are given out in 3.

**Comment by l0c4t0r** [comment]
> There are a lot of very larger TVL and volume errors. Also, when I look at the financial daily snapshots for TVL and daily volume in USD, there are times where the TVL looks normal, but the volume is way off - this is odd to me - though I am not sure if it requires an additional fix besides pricing.
> 
> https://subgraphs.messari.io/subgraph?endpoint=l0c4t0r/laboratory5&tab=protocol

Communicated with @steegecs on a call early. My fault for not including a link, the one that was reviewed was an older version that was synced. The subgraph based on code in this PR is still syncing as the pending version of l0c4t0r/laboratory5. The dashboard to QC should be:
https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/id/QmYTUWDYJapauCHZHR5FZiBk5Yte7KrzcMfjUFhuo4Zby1&tab=protocol

**Comment by l0c4t0r** [comment]
Latest commit:
1. Fix naming for reward tokens
2. Add mechanism to update fee rate for pools
3. Fix missing stakedOutputTokenAmounts

Ideally for 2, this should be triggered by the `setFee()` function on the PairFactory contract. Unfortunately, this function does not emit events and call handlers are not supported on optimism at the moment. 
Instead, the subgraph make contract calls to query the fee, and determine if an update is required.  This is currently set to be triggered by regular (but not overly frequent) events, such as PairCreated and DistributeRewards.

Due to the long sync times, I have deployed this version to l0c4t0r/laboratory4:
https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/id/QmVFoBtkbF3TrrJUXjLXV2oTbci9tVLTNk6jcp6fVMBxuS&tab=protocol

File: subgraphs/velodrome-finance/src/common/getters.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (l0c4t0r) [reply subgraphs/velodrome-finance/src/common/getters.ts]: should be fixed in e6e561c6114f88cb23eb7d779cef159c2f96fd83

File: subgraphs/velodrome-finance/src/common/getters.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (l0c4t0r) [reply subgraphs/velodrome-finance/src/common/getters.ts:287]: e6e561c6114f88cb23eb7d779cef159c2f96fd83 implements checks for updates to pool fees

[the remark this answers is not in the document]
Comment (steegecs) [reply subgraphs/velodrome-finance/src/common/getters.ts:287]: You say call handlers are not supported on Optimism. Will we be missing out on fee changes then? If so, is there some other way we can track fee updates?

File: subgraphs/velodrome-finance/src/common/metrics.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (steegecs) [reply subgraphs/velodrome-finance/src/common/metrics.ts:66]: Ah that makes sense! Thank you for explaining this.

**Comment by l0c4t0r** [comment]
Yes that makes sense, should've thought of it earlier!
I've put in `FEE_CHECK_INTERVAL_BLOCKS` in constants.ts to control this.  I've set this value to 900 which should translate to ~ every 30 minutes on optimism (Optimism has ~ 1 block every 2 seconds).
I also refactored `updateAllPoolFees` to be a bit more efficient by separating the list for stable pools and volatile pools.

I've set this to trigger on the Swap event, as well as the PairCreated event.  For the PairCreated event, the block number is ignored to ensure we are creating pools with the correct initial fee rate.

File: subgraphs/velodrome-finance/src/common/getters.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (l0c4t0r) [reply subgraphs/velodrome-finance/src/common/getters.ts:287]: Addressed below: https://github.com/messari/subgraphs/pull/881#issuecomment-1250003435

**Review by steegecs — DISMISSED** [review]
Alright. Looks good to me! Really nice work on this :)

**Comment by l0c4t0r** [comment]
Looks like I invalidated the review with a rebase to include a 1-line change from #1021 😅
Sorry @steegecs do you mind approving again?

**Review by steegecs — APPROVED** [review]
All good! Here is your approval.

**Comment by l0c4t0r** [comment]
Seems like there is a big (unrealistic) spike in revenue on 8/1 that's originating from the STG/USDC pool (0x7892d498f22c9393385f4a798788126f7090d2cc)
I am investigating now.

**Comment by steegecs** [comment]
> 

Sounds good. Let me know what you find!

**Comment by l0c4t0r** [comment]
The culprit of the incorrect pricing was due to this swap:
https://optimistic.etherscan.io/tx/0x7a74ac1a797054dc6cf6ed0c85b6dcf79e51fde751f7e1094ffe5447111eef2a

I added a minimum liquidity threshold of 1000 USD. Pricing from pools below this will be ignored.
This is deployed for QC at l0c4t0r/laboratory3: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/id/QmRBc5QKYyCRb2kNQCWvVw9AExkcSbVu8Zwn4LMcnDhkoe&tab=protocol

**Pull request merged**